core/services/aws/bloodDonation/acceptDonationRequest.ts

```diff
@@ -27,106 +27,216 @@ import {
 import BloodDonationDynamoDbOperations from '../commons/ddb/BloodDonationDynamoDbOperations'
 import {
   BloodDonationNotificationDTO,
-  NotificationStatus,
   NotificationType
 } from '../../../../commons/dto/NotificationDTO'
 import NotificationDynamoDbOperations from '../commons/ddb/NotificationDynamoDbOperations'
 import DonationNotificationModel, {
   BloodDonationNotificationFields
 } from '../../../application/models/dbModels/DonationNotificationModel'
 import AcceptedDonationDynamoDbOperations from '../commons/ddb/AcceptedDonationDynamoDbOperations'
+import { DonationNotificationAttributes } from '../../../application/notificationWorkflow/Types'
 
 const bloodDonationService = new BloodDonationService()
 const acceptDonationService = new AcceptDonationService()
 const userService = new UserService()
 const notificationService = new NotificationService()
 
-async function acceptDonationRequestLambda(
+async function acceptDonationRequest(
   event: AcceptDonationRequestAttributes
 ): Promise<APIGatewayProxyResult> {
   try {
-    const userProfile = await userService.getUser(
-      event.donorId,
-      new DynamoDbTableOperations<UserDetailsDTO, UserFields, UserModel>(new UserModel())
-    )
-    if (userProfile === null) {
-      throw new Error('Cannot find user')
+    const { donorId, seekerId, requestPostId, createdAt, status } = event
+
+    if (status !== AcceptDonationStatus.ACCEPTED && status !== AcceptDonationStatus.IGNORED) {
+      throw new Error('Invalid status for accept donation request')
     }
-    const donationPost = await bloodDonationService.getDonationRequest(
-      event.seekerId,
-      event.requestPostId,
-      event.createdAt,
-      new BloodDonationDynamoDbOperations<DonationDTO, DonationFields, BloodDonationModel>(
-        new BloodDonationModel()
-      )
+    const acceptanceRecord = await acceptDonationService.getAcceptanceRecord(
+      seekerId,
+      requestPostId,
+      donorId,
+      new AcceptedDonationDynamoDbOperations<
+      AcceptedDonationDTO,
+      AcceptedDonationFields,
+      AcceptDonationRequestModel
+      >(new AcceptDonationRequestModel())
     )
-    if (donationPost.status !== DonationStatus.PENDING) {
-      throw new Error('Donation request is no longer available for acceptance.')
+    if (acceptanceRecord !== null && acceptanceRecord.status === AcceptDonationStatus.COMPLETED) {
+      throw new Error('You already donated.')
+    }
+
+    if ((status === AcceptDonationStatus.ACCEPTED && acceptanceRecord !== null)) {
+      throw new Error('You already accepted the donation')
```

Comment:
i am thinking if we should throw errors here? idempotency dictates that the same request should have the same output.

Reply:
Updated in d1a75e62dc44681507dbd3b131846c6a53ab9363

Reply:
i think the message should be different though? it should still say, he this alrady exists but we just uswe 200

Reply:
got it. will update in next pr